warehouse/postgres/postgres.go

```diff
@@ -0,0 +1,520 @@
+package postgres
+
+import (
+	"compress/gzip"
+	"database/sql"
+	"encoding/csv"
+	"errors"
+	"fmt"
+	"github.com/lib/pq"
+	"github.com/rudderlabs/rudder-server/config"
+	"github.com/rudderlabs/rudder-server/rruntime"
+	"github.com/rudderlabs/rudder-server/services/filemanager"
+	"github.com/rudderlabs/rudder-server/services/stats"
+	"github.com/rudderlabs/rudder-server/utils/logger"
+	"github.com/rudderlabs/rudder-server/utils/misc"
+	warehouseutils "github.com/rudderlabs/rudder-server/warehouse/utils"
+	uuid "github.com/satori/go.uuid"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+)
+
+var (
+	warehouseUploadsTable string
+	stagingTablePrefix    string
+	maxParallelLoads      int
+)
+
+const (
+	host     = "host"
+	dbName   = "database"
+	user     = "user"
+	password = "password"
+	port     = "port"
+	sslMode  = "sslMode"
+)
+
+var dataTypesMap = map[string]string{
+	"int":      "bigint",
+	"float":    "numeric",
+	"string":   "text",
+	"datetime": "timestamp with time zone",
+	"boolean":  "boolean",
+}
+
+type HandleT struct {
+	DbHandle      *sql.DB
+	Db            *sql.DB
+	Namespace     string
+	CurrentSchema map[string]map[string]string
+	Warehouse     warehouseutils.WarehouseT
+	Upload        warehouseutils.UploadT
+	CloudProvider string
+	ObjectStorage string
+}
+
+type credentialsT struct {
+	host       string
+	dbName     string
+	user       string
+	password   string
+	schemaName string
+	port       string
+	sslMode    string
+}
+
+type optionalCredsT struct {
+	schemaName string
+}
+
+var primaryKeyMap = map[string]string{
+	"users":                      "id",
+	"identifies":                 "id",
+	warehouseutils.DiscardsTable: "row_id",
+}
+var partitionKeyMap = map[string]string{
+	"users":                      "id",
+	"identifies":                 "id",
+	warehouseutils.DiscardsTable: "row_id, column_name, table_name",
+}
+
+func connect(cred credentialsT) (*sql.DB, error) {
+	url := fmt.Sprintf("user=%v password=%v host=%v port=%v dbname=%v sslmode=%v",
```

Comment:
what are the values we are taking for sslmode from UI config. Are we supporting all these
![image](https://user-images.githubusercontent.com/8225792/82598934-0dc57e80-9bc9-11ea-8ef5-b5258c7d67fd.png)

Reply:
yes, i took the above image as a ref and added options in web app

Reply:
Let's add restriction in frontend/config backend destination configuration to allow only `disable` and `require` for sslmode. Refer https://github.com/lib/pq/blob/913c354da066cd5cee4687ed1ac3882fc2a9d366/doc.go#L65 for valid sslmode values for lib/pq. We don't have support for verify-ca and verify-full as of now, because we will need pass ssl certs for these modes to work.